myst_nb/sphinx_.py

```diff
@@ -0,0 +1,756 @@
+                # and replace the mime_bundle with the format specific output
+                # in a post-transform (run per output format on the cached AST)
+
+                # TODO how to output MyST Markdown?
+                # currently text/markdown is set to be rendered as CommonMark only,
+                # with headings dissallowed,
+                # to avoid "side effects" if the mime is discarded but contained
+                # targets, etc, and because we can't parse headings within containers.
+                # perhaps we could have a config option to allow this?
+                # - for non-commonmark, the text/markdown would always be considered
+                #   the top priority, and all other mime types would be ignored.
+                # - for headings, we would also need to parsing the markdown
+                #   at the "top-level", i.e. not nested in container(s)
+
+                figure_options = None
+                with suppress(KeyError):
+                    figure_options = self.get_cell_render_config(
+                        metadata, "figure", has_nb_key=False
+                    )
+
+                with create_figure_context(self, figure_options, line):
+                    mime_bundle = nodes.container(nb_element="mime_bundle")
+                    with self.current_node_context(mime_bundle):
+                        for mime_type, data in output["data"].items():
+                            mime_container = nodes.container(mime_type=mime_type)
+                            with self.current_node_context(mime_container):
+                                _nodes = self.nb_renderer.render_mime_type(
+                                    MimeData(
+                                        mime_type,
+                                        data,
+                                        cell_metadata=metadata,
+                                        output_metadata=output.get("metadata", {}),
+                                        cell_index=cell_index,
+                                        output_index=output_index,
+                                        line=line,
+                                    )
+                                )
+                                self.current_node.extend(_nodes)
+                            if mime_container.children:
+                                self.current_node.append(mime_container)
+                    if mime_bundle.children:
+                        self.add_line_and_source_path_r([mime_bundle], token)
+                        self.current_node.append(mime_bundle)
+            else:
+                self.create_warning(
+                    f"Unsupported output type: {output.output_type}",
+                    line=line,
+                    append_to=self.current_node,
+                    wtype=DEFAULT_LOG_TYPE,
+                    subtype="output_type",
+                )
+
+
+class SelectMimeType(SphinxPostTransform):
```

Comment:
@akhmerov this is where the mime type is selected

Reply:
Thanks. If I understand correctly, this way most of the sphinx machinery sees all the output nodes in the `ast`, until this `post-transform` is encountered. From what I understand so far, this is indeed preferred, and `jupyter-sphinx` uses the same approach.

As a quick note for your consideration: there's an [alternative approach](https://github.com/jupyter/jupyter-sphinx/blob/master/jupyter_sphinx/ast.py#L320-L342) to the post-transform that `jupyter-sphinx` uses. Here, instead of introducing a transform I create a node that presents itself as its highest priority child to the builder. I don't think the differences between the two approaches are important, so I'm mainly bringing this up for completeness.

Reply:
👍 yep cheers, as you say not too much difference, but I felt the post-transform was slightly more explicit and more of a "standard" sphinx pattern